controllers/cinder_controller.go

```diff
@@ -630,31 +629,50 @@ func (r *CinderReconciler) reconcileNormal(ctx context.Context, instance *cinder
 		instance.Status.Conditions.Set(c)
 	}
 
-	// deploy cinder-backup
-	cinderBackup, op, err := r.backupDeploymentCreateOrUpdate(ctx, instance)
-	if err != nil {
-		instance.Status.Conditions.Set(condition.FalseCondition(
-			cinderv1beta1.CinderBackupReadyCondition,
-			condition.ErrorReason,
-			condition.SeverityWarning,
-			cinderv1beta1.CinderBackupReadyErrorMessage,
-			err.Error()))
-		return ctrl.Result{}, err
-	}
-	if op != controllerutil.OperationResultNone {
-		r.Log.Info(fmt.Sprintf("Deployment %s successfully reconciled - operation: %s", instance.Name, string(op)))
-	}
+	// deploy cinder-backup, but only if necessary
+	//
+	// Many OpenStack deployments don't use the cinder-backup service (it's optional),
+	// so there's no need to deploy it unless it's required.
+	var backupCondition *condition.Condition
+	if instance.Spec.CinderBackup.Replicas > 0 {
+		cinderBackup, op, err := r.backupDeploymentCreateOrUpdate(ctx, instance)
+		if err != nil {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				cinderv1beta1.CinderBackupReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				cinderv1beta1.CinderBackupReadyErrorMessage,
+				err.Error()))
+			return ctrl.Result{}, err
+		}
+		if op != controllerutil.OperationResultNone {
+			r.Log.Info(fmt.Sprintf("Deployment %s successfully reconciled - operation: %s", instance.Name, string(op)))
+		}
 
-	// Mirror CinderBackup status' ReadyCount to this parent CR
-	instance.Status.CinderBackupReadyCount = cinderBackup.Status.ReadyCount
+		// Mirror CinderBackup status' ReadyCount to this parent CR
+		instance.Status.CinderBackupReadyCount = cinderBackup.Status.ReadyCount
 
-	// Mirror CinderBackup's condition status
-	c = cinderBackup.Status.Conditions.Mirror(cinderv1beta1.CinderBackupReadyCondition)
-	if c != nil {
-		instance.Status.Conditions.Set(c)
+		// Mirror CinderBackup's condition status
+		backupCondition = cinderBackup.Status.Conditions.Mirror(cinderv1beta1.CinderBackupReadyCondition)
+
+	} else {
+		// Clean up cinder-backup if there are no replicas
+		err = r.backupCleanupDeployment(ctx, instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if backupCondition != nil {
+		// If there's a backupCondition then set that as the CinderBackupReadyCondition
+		instance.Status.Conditions.Set(backupCondition)
+	} else {
+		// The CinderBackup is ready, even if the service wasn't deployed.
+		// Using "condition.DeploymentReadyMessage" here because that is what gets mirrored
+		// as the message for the other Cinder children when they are successfully-deployed
+		instance.Status.Conditions.MarkTrue(cinderv1beta1.CinderBackupReadyCondition, condition.DeploymentReadyMessage)
```

Comment:
On initial deployment we can set it like this because there's nothing running, but when modifying a system that has cinder-backup present we need to wait until the CinderBackup has been actually removed, which should only happen once the statefulset and pods have actually been removed, right?

Reply:
My understanding is the CinderXxxReadyCondition(s) are used for the purpose of declaring the cinder service itself is ready, and generally that means the number of "ready" replicas is either >0 when replicas is 0, or when replicas *is* 0. For that reason I don't see when it's necessary to wait for the final backup replica to be deleted before declaring the service is "ready"